src/pebble.py

```diff
@@ -70,10 +76,30 @@ def change_config(self, container: ops.model.Container) -> None:
         try:
             synapse.execute_migrate_config(container=container, charm_state=self._charm_state)
             synapse.enable_metrics(container=container)
+            if self._charm_state.saml_config is not None:
+                logger.debug("pebble.change_config: Enabling SAML")
+                synapse.enable_saml(container=container, charm_state=self._charm_state)
```

Comment:
I believe that is not needed here because in this case the plan will be changed (when a configuration changes, the env vars change too). Do you see a different scenario?

Reply:
Hmm, somewhat dissatisfying that we have cases where we can rely on pebble to perform the restart and others where we have to be explicit. If we change something and pebble doesn't do the restart anymore, will we know that we have do to the restart? Perhaps at least a comment?

Reply:
It would be nice if this was controlled by the container.replan itself.
I agree that might be confusing in the future. I'll add a comment warning about it similar to the one in enable_saml.

Reply:
@benhoyt any comment on this?

Reply:
It looks to me like `synapse.execute_migrate_config`, `synapse.enable_metrics`, and `synapse.enable_saml` only do a `container.push`, not update the plan. If I'm reading that correctly, `container.replan()` won't have any effect -- you'll need to use `container.restart(service)` explicitly.

In general, Pebble doesn't monitor config files for changes, so there are two cases needed:

1) The plan changing, e.g., the service's environment variables or command line. In this case a Pebble `replan` will restart the service.
2) A config file changed but the plan hasn't changed. In this case Pebble will need to be explicitly told to restart the services in question.

Here it's hard to tell -- it's very round-about, because `self.replan(container)` does add a layer, which depends on `synapse.get_environment()` (which in turn depends on the charm state). So if the charm state has changed since the plan was last generated, a replan would be enough.

In this case, since config files are involved as well, I'd probably change this to unconditionally do a restart (or maybe there's a way tell Synapse to do a graceful, zero-downtime restart, like sending SIGHUP?), seeing it's going to happen sometimes anyway. However, the ideal would be to detect if any of the config files changed (during the pull-push sequences) and only restart if they or the plan has changed.

Reply:
I'm curious about the "In general, Pebble doesn't monitor config files for changes" part - do you know whether this is by design or perhaps Pebble will at some point actively monitor for config changes and act?

Reply:
In this case the charm would have to do a replan and restart everytime, right? Or only the restart would update the plan too?
Synapse supports SIGHUP only for specific configurations as mentioned [here](https://github.com/matrix-org/synapse/issues/16071).

Reply:
@merkata Re Pebble monitoring for config file changes. This is not so much "by design" as by it not having been designed yet. I wouldn't rule it out, though it's a non-trivial feature, so I wouldn't hold your breath. If you're interested, you could look at writing a spec for it and see if there's appetite for it?

@amandahla Yep, that's right - replan will restart only if there's a *plan* change.